File: services/user-service/src/main/java/com/visor/school/userservice/service/ParentService.java
package com.visor.school.userservice.service;

import java.util.List;
import java.util.UUID;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import com.visor.school.userservice.model.Parent;
import com.visor.school.userservice.model.Relationship;
import com.visor.school.userservice.repository.ParentRepository;

/**
 * Parent service for managing parent-student relationships
 */
@Service
@Transactional
public class ParentService {

    private final Logger logger = LoggerFactory.getLogger(getClass());
    private final ParentRepository parentRepository;

    public ParentService(ParentRepository parentRepository) {
        this.parentRepository = parentRepository;
    }

    /**
     * Link a student to a parent
     */
    public Parent linkStudentToParent(
        UUID userId,
        UUID studentId,
        Relationship relationship,
        boolean isPrimary
    ) {
        logger.info("Linking student {} to parent {} with relationship: {}", studentId, userId, relationship);

        if (parentRepository.existsByUserIdAndStudentId(userId, studentId)) {
            throw new IllegalArgumentException("Parent-student relationship already exists");
        }

        Parent parent = new Parent(
            userId,
            studentId,
            relationship,
            isPrimary
        );

        Parent saved = parentRepository.save(parent);
        logger.info("Parent-student relationship created: {}", saved.getId());

        return saved;
    }

    /**
     * Get all children (students) for a parent
     */
    @Transactional(readOnly = true)
    public List<Parent> getChildrenByParent(UUID userId) {
        return parentRepository.findByUserId(userId);
    }

    /**
     * Get all parents for a student
     */
    @Transactional(readOnly = true)
    public List<Parent> getParentsByStudent(UUID studentId) {
        return parentRepository.findByStudentId(studentId);
    }

    /**
     * Remove parent-student relationship
     */
    public void unlinkStudentFromParent(UUID userId, UUID studentId) {
        logger.info("Unlinking student {} from parent {}", studentId, userId);

        List<Parent> parents = parentRepository.findByUserId(userId);
        Parent parent = parents.stream()
            .filter(p -> p.getStudentId().equals(studentId))
            .findFirst()
            .orElseThrow(() -> new IllegalArgumentException("Parent-student relationship not found"));

        parentRepository.delete(parent);
        logger.info("Parent-student relationship removed");
    }
}

File: services/user-service/src/main/kotlin/com/visor/school/userservice/service/ParentService.kt
package com.visor.school.userservice.service

import com.visor.school.userservice.model.Parent
import com.visor.school.userservice.model.Relationship
import com.visor.school.userservice.repository.ParentRepository
import org.slf4j.LoggerFactory
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional
import java.util.UUID

/**
 * Parent service for managing parent-student relationships
 */
@Service
@Transactional
class ParentService(
    private val parentRepository: ParentRepository
) {
    private val logger = LoggerFactory.getLogger(ParentService::class.java)

    /**
     * Link a student to a parent
     */
    fun linkStudentToParent(
        userId: UUID,
        studentId: UUID,
        relationship: Relationship,
        isPrimary: Boolean = false
    ): Parent {
        logger.info("Linking student $studentId to parent $userId with relationship: $relationship")

        if (parentRepository.existsByUserIdAndStudentId(userId, studentId)) {
            throw IllegalArgumentException("Parent-student relationship already exists")
        }

        val parent = Parent(
            userId = userId,
            studentId = studentId,
            relationship = relationship,
            isPrimary = isPrimary
        )

        val saved = parentRepository.save(parent)
        logger.info("Parent-student relationship created: ${saved.id}")

        return saved
    }

    /**
     * Get all children (students) for a parent
     */
    @Transactional(readOnly = true)
    fun getChildrenByParent(userId: UUID): List<Parent> {
        return parentRepository.findByUserId(userId)
    }

    /**
     * Get all parents for a student
     */
    @Transactional(readOnly = true)
    fun getParentsByStudent(studentId: UUID): List<Parent> {
        return parentRepository.findByStudentId(studentId)
    }

    /**
     * Remove parent-student relationship
     */
    fun unlinkStudentFromParent(userId: UUID, studentId: UUID) {
        logger.info("Unlinking student $studentId from parent $userId")

        val parent = parentRepository.findByUserId(userId)
            .find { it.studentId == studentId }
            ?: throw IllegalArgumentException("Parent-student relationship not found")

        parentRepository.delete(parent)
        logger.info("Parent-student relationship removed")
    }
}

